### PR TITLE
Add `MongoDocumentStore` 

### PR DESCRIPTION
**Pull request by Disiok**
### Summary
* Add MongoDB as `DocumentStore` backend

### Notes
* Right now, we still store additional metadata (e.g. index struct) as json blobs on disk. Future PR to unify storage layer across nodes, index structs, and other configuration related metadata

### Todos
- [x] Fix support for graph
- [x] Fix support for document hash
- [x] Unit testing
- [x] Notebooks
- [x] Lint

### CODE REVIEW

File: gpt_index/docstore/__init__.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from gpt_index.docstore.types import DocumentStore
```
Comment (jerryjliu) [inline gpt_index/docstore/__init__.py:1]: i'd name this BaseDocumentStore, and keep DocumentStore for backwards compat with SimpleDocumentStore

**Review by jerryjliu — COMMENTED** [review]
looks pretty good! love the tests. one general comment about the naming

A few sections where this could be good to add - as a section in "Key Components" in the docs, also in the API reference

File: gpt_index/docstore/__init__.py (hunk 1 above)
Comment (Disiok) [reply gpt_index/docstore/__init__.py]: Revised.

**Review by jerryjliu — APPROVED** [review]
looks good to me!

**Pull request merged**